rosbag2/src/rosbag2/sequential_reader.cpp

```diff
@@ -71,9 +76,29 @@ SequentialReader::open(
   }
 }
 
+std::string SequentialReader::get_next_file()
+{
+  assert(current_file_iterator_ != file_paths_.end());
+  current_file_iterator_++;
+  return *current_file_iterator_;
+}
+
+bool SequentialReader::has_next_file() const
+{
+  return current_file_iterator_ + 1 != file_paths_.end();
+}
+
 bool SequentialReader::has_next()
 {
   if (storage_) {
+    // If there's no new message, check if there's at least another file to read and update storage
+    // to read from there. Otherwise, check if there's another message.
+    if (!storage_->has_next()) {
+      if (has_next_file()) {
+        const auto current_file = get_next_file();
```

Comment:
this line is the motivation for my previous comments. At this point here it's not clear to me without having insight knowledge of the code, that the `current_file_iterator_` has been moved forward and thus `open_read_only` can easily be called without calling `open` on this class again.